src/compiler/checker.ts

```diff
@@ -14400,12 +14432,126 @@ namespace ts {
             }
         }
 
+        function checkUnusedLocals(node: FunctionDeclaration | MethodDeclaration | ConstructorDeclaration | FunctionExpression | ArrowFunction | ForInStatement | Block): void {
+            checkUnusedIdentifiers(node);
+            checkUnusedParameters(node);
+        }
+
+        function checkUnusedIdentifiers(node: FunctionDeclaration | MethodDeclaration | ConstructorDeclaration | FunctionExpression | ArrowFunction | ForInStatement | Block): void {
+            if (compilerOptions.noUnusedLocals && !isInAmbientContext(node)) {
+                for (const key in node.locals) {
+                    if (hasProperty(node.locals, key)) {
+                        const local = node.locals[key];
+                        if (!local.hasReference && local.valueDeclaration && local.valueDeclaration.kind) {
+                            if (local.valueDeclaration.kind !== SyntaxKind.Parameter) {
+                                error(local.valueDeclaration, Diagnostics._0_is_declared_but_never_used, key);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        function checkUnusedParameters(node: FunctionDeclaration | MethodDeclaration | ConstructorDeclaration | FunctionExpression | ArrowFunction | ForInStatement | Block): void {

```

Comment:
can we combine these two functions in one. seems like we are looping over the same set of symbols twice. just loop once, and do the check on option before you report the error.